crates/bitcoind_rpc/src/lib.rs

```diff
@@ -127,13 +127,45 @@ impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
     }
 
     /// Emit the next block height and header (if any).
-    pub fn next_header(&mut self) -> Result<Option<(u32, Header)>, bitcoincore_rpc::Error> {
-        poll(self, |hash| self.client.get_block_header(hash))
+    pub fn next_header(&mut self) -> Result<Option<Emission<Header>>, bitcoincore_rpc::Error> {
+        Ok(poll(self, |hash| self.client.get_block_header(hash))?
+            .map(|(checkpoint, block)| Emission { block, checkpoint }))
     }
 
     /// Emit the next block height and block (if any).
-    pub fn next_block(&mut self) -> Result<Option<(u32, Block)>, bitcoincore_rpc::Error> {
-        poll(self, |hash| self.client.get_block(hash))
+    pub fn next_block(&mut self) -> Result<Option<Emission<Block>>, bitcoincore_rpc::Error> {
+        Ok(poll(self, |hash| self.client.get_block(hash))?
+            .map(|(checkpoint, block)| Emission { block, checkpoint }))
+    }
+}
+
+/// An emission from [`Emitter`].
+#[derive(Debug)]
+pub struct Emission<B> {
+    /// Either a full [`Block`] or [`Header`].
+    pub block: B,
+    /// The checkpoint of the emitted block/header.
+    pub checkpoint: CheckPoint,
+}
+
+impl<B> Emission<B> {
+    /// The block height of this emitted block.
+    pub fn block_height(&self) -> u32 {
+        self.checkpoint.height()
+    }
+
+    /// The block hash of this emitted block.
+    pub fn block_hash(&self) -> BlockHash {
+        self.checkpoint.hash()
+    }
+
+    /// The [`BlockId`] of where this emitted block should connect to.
+    pub fn connected_to(&self) -> BlockId {
+        match self.checkpoint.prev() {
+            Some(prev_cp) => prev_cp.block_id(),
+            // there is no previous checkpoint, so just connect with itself
+            None => self.checkpoint.block_id(),
```

Comment:
Why isn't it connected to the block's prev block hash block? This should never emit the genesis block right so this will always work. The key thing I am confused about (which there is a high probability you've already explained it to me months ago) is why we even need this checkpoint thing in the emission.

Reply:
As discussed in the call, if `start_height` is > 1 and the original chain's tip is 0, the first emission will not connect to the prev block.

If the caller constructs `Emitter` from the tip of a `LocalChain`, this should not happen if the bitcoind genesis hash is the same as `LocalChain`'s genesis hash. However, we can't guarantee that the caller will do this. The alternative would be to return an error, but that just makes the API cumbersome.